### ISSUE TITLE
Previous device units not removed when renaming network interface

### ISSUE THREAD

**Issue by dechamps** [systemd/systemd#16967]
**systemd version the issue has been seen with**
246.4-1

**Used distribution**
Debian Unstable (kernel 5.7.0-3-amd64)

**Steps to reproduce the problem**
```
# ip link add testbr type bridge
# ip link set testbr name testbr2
# systemctl | grep testbr
```

**Expected result**
```
  sys-devices-virtual-net-testbr2.device   loaded active plugged   /sys/devices/virtual/net/testbr2
  sys-subsystem-net-devices-testbr2.device   loaded active plugged   /sys/subsystem/net/devices/testbr2
```

**Actual result**
```
  sys-devices-virtual-net-testbr.device   loaded active plugged   /sys/devices/virtual/net/testbr
  sys-devices-virtual-net-testbr2.device   loaded active plugged   /sys/devices/virtual/net/testbr2
  sys-subsystem-net-devices-testbr.device   loaded active plugged   /sys/subsystem/net/devices/testbr
  sys-subsystem-net-devices-testbr2.device   loaded active plugged   /sys/subsystem/net/devices/testbr2
```

If I keep renaming the interface, then device units will continue to accumulate as above.

Worse, if the interface is removed (`ip link del testbr2`), then only the device units for the current name are removed. It doesn't seem possible to get rid of the device units for the previous names, which seem to linger around forever.

**Comment by dechamps** [comment]
One workaround is to create a dummy interface with the old name and then remove it to trick systemd into noticing its absence:

```
# ip link add testbr type bridge
# ip link set testbr name testbr2
# systemctl | grep testbr
  sys-devices-virtual-net-testbr.device   loaded active plugged   /sys/devices/virtual/net/testbr
  sys-devices-virtual-net-testbr2.device   loaded active plugged   /sys/devices/virtual/net/testbr2
  sys-subsystem-net-devices-testbr.device   loaded active plugged   /sys/subsystem/net/devices/testbr
  sys-subsystem-net-devices-testbr2.device   loaded active plugged   /sys/subsystem/net/devices/testbr2
# ip link add testbr type bridge
# ip link del testbr
# systemctl | grep testbr                                      
  sys-devices-virtual-net-testbr2.device   loaded active plugged   /sys/devices/virtual/net/testbr2
  sys-subsystem-net-devices-testbr2.device   loaded active plugged   /sys/subsystem/net/devices/testbr2
```

**Comment by yuwata** [comment]
Note that `daemon-reload` removes non-existent .device under /sys/devices, but not under /sys/subsystem.
```
$ sudo ip link add hoge type dummy
$ sudo ip link set hoge name hoge2
$ systemctl | grep hoge
  sys-devices-virtual-net-hoge.device             loaded active plugged   /sys/devices/virtual/net/hoge
  sys-devices-virtual-net-hoge2.device            loaded active plugged   /sys/devices/virtual/net/hoge2
  sys-subsystem-net-devices-hoge.device           loaded active plugged   /sys/subsystem/net/devices/hoge
  sys-subsystem-net-devices-hoge2.device          loaded active plugged   /sys/subsystem/net/devices/hoge2
$ sudo systemctl daemon-reload
$ systemctl | grep hoge
  sys-devices-virtual-net-hoge2.device            loaded active plugged   /sys/devices/virtual/net/hoge2
  sys-subsystem-net-devices-hoge.device           loaded active plugged   /sys/subsystem/net/devices/hoge
  sys-subsystem-net-devices-hoge2.device          loaded active plugged   /sys/subsystem/net/devices/hoge2
```

**Comment by yuwata** [comment]
Hmm, after renaming, udev database has old alias:
```
$ cat n7
I:487922062
E:ID_NET_NAMING_SCHEME=v245
E:ID_NET_DRIVER=dummy
E:ID_NET_LINK_FILE=/etc/systemd/network/99-default.link
E:SYSTEMD_ALIAS=/sys/subsystem/net/devices/hoge /sys/subsystem/net/devices/hoge2
G:systemd
Q:systemd
```
Then, pid1 saids
```
systemd[1]: sys-subsystem-net-devices-hoge.device: Device sys-subsystem-net-devices-hoge.device appeared twice with different sysfs paths /sys/devices/virtual/net/hoge and /sys/devices/virtual/net/hoge2, ignoring the latter.
```

**Comment by yuwata** [comment]
Fix is waiting in #16968.

**Comment by dechamps** [comment]
I just realized that the workaround I suggested (create a new dummy interface with the old name, then delete it) only seems to work for virtual devices. For physical devices (e.g. `/sys/devices/pci…`) it doesn't work. I suspect that's because the dummy interface doesn't have the same device path as the real interface.
